test/e2e/e2e_test.go

```diff
@@ -0,0 +1,204 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"context"
+	"errors"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	metricsv1beta1 "k8s.io/metrics/pkg/apis/metrics/v1beta1"
+	metrics "k8s.io/metrics/pkg/client/clientset/versioned"
+)
+
+const ns = "prometheus-adapter-e2e"
+
+func initializeClient(t *testing.T) (clientset.Interface, metrics.Interface) {
+	kubeconfig := os.Getenv("KUBECONFIG")
+	if len(kubeconfig) == 0 {
+		t.Fatal("KUBECONFIG not provided")
+	}
+
+	cfg, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+	if err != nil {
+		t.Fatalf("Error during client configuration with %v", err)
+	}
+
+	clientSet, err := clientset.NewForConfig(cfg)
+	if err != nil {
+		t.Fatalf("Error during client creation with %v", err)
+	}
+
+	dynClient, err := dynamic.NewForConfig(cfg)
+	if err != nil {
+		t.Fatalf("Error during dynamic client creation with %v", err)
+	}
+
+	metricsClientSet, err := metrics.NewForConfig(cfg)
+	if err != nil {
+		t.Fatalf("Error during metrics client creation with %v", err)
+	}
+
+	ctx := context.Background()
+	err = waitForPrometheusReady(ctx, t, dynClient, ns, "prometheus")
+	if err != nil {
+		t.Fatalf("Prometheus instance 'prometheus' not ready: %v", err)
+	}
+	err = waitForDeploymentReady(ctx, t, clientSet, ns, "prometheus-adapter")
+	if err != nil {
+		t.Fatalf("Deployment prometheus-adapter not ready: %v", err)
+	}
+	return clientSet, metricsClientSet
+}
+
+func waitForPrometheusReady(ctx context.Context, t *testing.T, client dynamic.Interface, namespace string, name string) error {
```

Comment:
what do you think about reusing the prometheus-operator e2e test framework for that? https://github.com/prometheus-operator/prometheus-operator/blob/1dfecab7c2706f3591d6bb6d5eb8f6b2518f1b45/test/framework/prometheus.go#L309-L320

Reply:
That's a good idea. I tried but got stuck in dependency hells, as prometheus-operator brings Kubernetes, prometheus and opentelemetry dependencies in different versions than prometheus-adapter.
However, I could use [prometheus-operator client](https://pkg.go.dev/github.com/prometheus-operator/prometheus-operator/pkg/client) (which is a different module with fewer dependencies) to make this function less verbose, if you think it's better.

Reply:
> That's a good idea. I tried but got stuck in dependency hells, as prometheus-operator brings Kubernetes, prometheus and opentelemetry dependencies in different versions than prometheus-adapter.

Oh that's unfortunate, I am fine with the current approach, just though that it could maybe be simplified by importing existing code.

> However, I could use [prometheus-operator client](https://pkg.go.dev/github.com/prometheus-operator/prometheus-operator/pkg/client) (which is a different module with fewer dependencies) to make this function less verbose, if you think it's better.

Yeah that's a good idea

Reply:
I refactored the tests to use the prometheus-operator client.